text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+However, instead of a `buildpack.toml` file, hooks have a `hook.toml` file:
+```toml
+api = "<buildpack API version>"
+
+[hook]
+id = "<hook ID>"
+name = "<hook name>"
+version = "<hook version>"
+homepage = "<hook homepage>"
+description = "<hook description>"
+keywords = [ "<string>" ]
+
+[[hook.licenses]]
+type = "<string>"
+uri = "<uri>"
+```
+
+Hooks may be packaged and examined similar to buildpacks, but with analogous `pack hook` subcommands.
+
+Other pack CLI commands, such as `pack builder create`, will be extended to include support for hooks.
+
+Unlike buildpacks,
+- Hooks must not be included in a meta-buildpacks
+- Hooks must not have `order`/`group` definitions in `hook.toml`
+
+Hooks participate in the buildpack detection process, with the same interface for `/bin/detect`.
+However,
+- `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not presetn.
+- Hooks may only output `provides` entries to the build plan. They must not output `requires`.
+- Hooks must all proceed regular buildpacks in `order` definitions (e.g., in `builder.toml`).
+- Hooks are always `optional`.
+
+Hooks generate Dockerfiles before the regular buildpack build phase.
+To generate these Dockerfiles, the lifecycle executes the hook's `/bin/build` executable with the same interface as regular buildpacks.
+However,
+- Hooks `/bin/build` must not write to the app directory.
+- Hooks `/bin/build` may be executed in parallel.
+- Hooks `<layers>` directory is replaced by an `<output>` directory.
+- If a hook is missing `/bin/build`, the hook root is treated as a pre-populated `<output>` directory.
+
+After `/bin/build` executes, the `<output>` directory may contain
+- `build.toml`, with the same contents as a normal buildpack's `build.toml`, but
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `build.Dockerfile` or `Dockerfile`
+- `launch.toml`, with the same contents as a normal buildpack's `launch.toml`, but
+  - Without the `processes` table array
+  - Without the `slices` table array
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `run.Dockerfile` or `Dockerfile`
+
+- Either `Dockerfile` or either or both of `build.Dockerfile` and `run.Dockerfile`
+
+Support for other instruction formats, e.g., LLB JSON files, could be added in the future.
+
+`build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively.
+
+If no Dockerfiles are present, `/bin/build` may still consume build plan entries and add metadata to `build.toml`/`launch.toml`.
+
+Dockerfiles are applied to their corresponding base images after all hooks are executed and before any regular buildpacks are executed.
+Dockerfiles are applied in the order determined during buildpack detection.
+
+All Dockerfiles are provided with `base_image` and `build_id` args.
```

Comment:
Could you provide a more concrete use case for how `build_id` could be used? Who is able to provide this value to invalidate the cache?

Reply:
The lifecycle provides it automatically.
Example is here: https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R180